docs/docs/creating-a-source-plugin.md

```diff
@@ -18,147 +18,426 @@ At a high-level, a source plugin:
+For Gatsby to automatically infer a relationship, you need to create a field called `author___NODE` on the Post object to hold the relationship to Authors before you create the node. The value of this field should be the node ID of the Author.
 
-When an object node is deleted, Gatsby _does not_ delete any referenced entities. When using foreign-key references, it's a source plugin's responsibility to clean up any dangling entity references.
+```javascript:title=source-plugin/gatsby-node.js
+exports.sourceNodes = ({ actions, createContentDigest }) => {
+  const { createNode } = actions
+  createNode({
+    // Data for the Post node
+    author___NODE: `<the-authors-gatsby-node-id>`, // highlight-line
+    // Required fields
+    id: `a-node-id`,
+    parent: null
+    children: [],
+    internal: {
+      type: `post`,
+      contentDigest: createContentDigest(fieldData),
+    }
+  })
+}
+```
 
-##### Creating the relationship
+For a stricter GraphQL schema, you can specify the exact field and value to link nodes with using schema customization APIs.
 
-Suppose you want to create a relationship between Posts and Authors, and you want to call the field `author`.
+```javascript:title=source-plugin/gatsby-node.js
+exports.sourceNodes = ({ actions, createContentDigest }) => {
+  const { createNode } = actions
+  createNode({
+    // Data for the Post node
+    // highlight-start
+    author: {
+      name: `Jay Gatsby`,
+    },
+    // highlight-end
+    // Required fields
+    id: `a-node-id`,
+    parent: null
+    children: [],
+    internal: {
+      type: `post`,
+      contentDigest: createContentDigest(fieldData),
+    }
+  })
+}
 
-Before you pass the Post object and Author object into `createNode` and create the respective nodes, you need to create a field called `author___NODE` on the Post object to hold the relationship to Authors. The value of this field should be the node ID of the Author.
+exports.createSchemaCustomization = ({ actions }) => {
+  const { createTypes } = actions
+  createTypes(`
+    type Post implements Node {
+      id: ID!
+      # create a relationship between Post and the File nodes for optimized images
+      author: Author @link(from: "author.name" by: "name") // highlight-line
+      # ... other fields
+    }`)
+}
+```
 
 ##### Creating the reverse relationship
 
 It's often convenient for querying to add to the schema backwards references. For example, you might want to query the Author of a Post but you might also want to query all the posts an author has written.
 
-If you want to call this field on `Author` `posts`, you would create a field called `posts___NODE` to hold the relationship to Posts. The value of this field should be an array of Post IDs.
+If you want to call this field on `Author` `posts` using the inference method, you would create a field called `posts___NODE` to hold the relationship to Posts. The value of this field should be an array of Post IDs.
```

Comment:
Posts vs. `posts` here is a bit confusing. Can we be consistent?

Reply:
You're right this is confusing. I tried to update it to use the capital with backticks when it's referring to the GraphQL node type and to a lowercased posts when it's used to refer to posts more colloquially. How does it look now?